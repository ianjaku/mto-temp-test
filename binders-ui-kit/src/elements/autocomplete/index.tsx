import * as React from "react";
import Chip from "@material-ui/core/Chip";
import { ChipComponentProps } from "./contract";
import CircularProgress from "../circularprogress";
import ClearIcon from "@material-ui/icons/Clear";
import { TranslationKeys as TK } from "@binders/client/lib/react/i18n/translations";
import Tooltip from "@material-ui/core/Tooltip";
import autobind from "class-autobind";
import colors from "../../variables";
import cx from "classnames";
import { withTranslation } from "@binders/client/lib/react/i18n";
import "./autocomplete.styl";

export interface IAutocompleteItem {
    label: string;
    rawLabel: string;
    id: string;
    value: string;
    isNew?: boolean;
}

enum LoadingState {
    Loading = "Loading",
    Initial = "Initial",
    Complete = "Complete",
}

export interface IAutocompleteProps {
    data: IAutocompleteItem[];
    onUpdateSearchTerm?: (term: string) => void;
    // eslint-disable-next-line @typescript-eslint/ban-types
    onAddNewClick: (item: object) => void;
    // eslint-disable-next-line @typescript-eslint/ban-types
    onDeleteChipClick: (item: object) => void;
    selectedItems: IAutocompleteItem[];
    totalResults?: number;
    maxAllowedSelectedItems?: number;
    ChipComponent?: React.FC<ChipComponentProps> | React.ComponentClass<ChipComponentProps>;
    wrapChips?: boolean;
    saveOnBlur?: boolean;
    placeholder?: string;
    onInputRef?: (ref: HTMLTextAreaElement) => void;
    isLoading?: boolean;
    renderLeadingSlot?: () => React.ReactNode;
    renderTrailingSlot?: () => React.ReactNode;
    allowItemCreation?: boolean;
    noMatchesOverride?: string;
    hideIcon?: boolean;
}

export interface IAutocompleteState {
    searchTerm: string;
    loadingState?: LoadingState;
}

class Autocomplete extends React.Component<IAutocompleteProps, IAutocompleteState> {

    private autocomplete = null;
    private t;

    constructor(props) {
        super(props);
        this.t = props.t;
        autobind(this);
        this.state = {
            searchTerm: "",
            loadingState: LoadingState.Initial,
        };
    }

    public componentDidUpdate(prevProps: IAutocompleteProps) {
        const { isLoading: wasLoading } = prevProps;
        if (!wasLoading && this.props.isLoading) {
            this.setState({
                loadingState: LoadingState.Loading,
            });
        }
        if (wasLoading && !this.props.isLoading) {
            this.setState({
                loadingState: LoadingState.Complete,
            });
        }
    }

    private maybeAddMoreArea() {
        const { selectedItems, maxAllowedSelectedItems, placeholder, onInputRef } = this.props;

        if (selectedItems.length >= maxAllowedSelectedItems) {
            return null;
        }

        const isClean = selectedItems.length === 0;
        const plHolder = isClean ?
            (placeholder || this.t(TK.User_AddUserOrUsergroup)) :
            this.t(TK.General_AddMore);
        return (
            <textarea
                rows={1}
                placeholder={plHolder}
                onChange={this.onChange}
                onKeyPress={this.onRequestAdd}
                onBlur={this.onBlur}
                className="autocomplete-textarea"
                value={this.state.searchTerm}
                ref={onInputRef}
            />
        );
    }

    public render() {
        const { selectedItems, wrapChips, renderLeadingSlot, renderTrailingSlot, hideIcon } = this.props;
        const isClean = selectedItems.length === 0;
        const className = cx("autocomplete", {
            "autocomplete--is-dirty": !isClean,
            "autocomplete--hideIcon": hideIcon || !!renderLeadingSlot,
            "autocomplete--loadingComplete": this.state.loadingState === LoadingState.Complete,
        });
        return (
            <div className="autocomplete-wrapper">
                <div
                    ref={autocomplete => { this.autocomplete = autocomplete; }}
                    className={className}
                >
                    {renderLeadingSlot && renderLeadingSlot()}
                    <div className="autocomplete-pane" style={wrapChips ? { flexFlow: "wrap" } : {}}>
                        <div className="autocomplete-pane-chips">
                            {this.renderChips()}
                            {this.maybeAddMoreArea()}
                        </div>
                    </div>
                    {renderTrailingSlot && renderTrailingSlot()}
                    {this.props.isLoading && (
                        <div className="autocomplete-alignEnd">
                            {CircularProgress("", {}, undefined, colors.disabledColor)}
                        </div>
                    )}
                </div>
                {this.renderPrompts()}
            </div>
        );
    }

    private onChange({ target: { value } }) {
        this.setState({
            searchTerm: value,
            loadingState: LoadingState.Initial,
        });
        if (this.props.onUpdateSearchTerm) {
            this.props.onUpdateSearchTerm(value);
        }
    }

    /* RENDERS */

    private renderChip(label: string, value: string, isNew: boolean, isUser: boolean) {
        const { ChipComponent, maxAllowedSelectedItems } = this.props;
        const usesCustomChip = !!ChipComponent;
        const ChipToRender = this.props.ChipComponent || Chip;
        return (
            <div className={cx("autocomplete-chipWrapper", { "autocomplete-chipWrapper--singleChip": maxAllowedSelectedItems === 1 })}>
                <ChipToRender
                    classes={{
                        root: `autocomplete-chip ${isUser ? "autocomplete-chip--user" : "autocomplete-chip--group"}`,
                    }}
                    className={"autocomplete-chip"}
                    deleteIcon={<ClearIcon style={{ fill: colors.disabledColor, marginTop: "1px", width: "14px" }} />}
                    onDelete={this.onDeleteChip(value, label, isNew)}
                    label={label}
                    value={value}
                    {...(usesCustomChip ? { isNew } : {})}
                />
            </div>
        );
    }

    private renderChips() {
        const { selectedItems } = this.props;

        return selectedItems.map(({ id, label, value, isNew = false }) => {
            const tooltipProps = isNew ?
                { title: this.t(TK.User_NoUserOrUserGroup) } :
                { title: "", open: false }
            return (
                <Tooltip key={value} {...tooltipProps}>
                    <>
                        {this.renderChip(label, value, isNew, id.startsWith("uid-"))}
                    </>
                </Tooltip>
            );
        });
    }

    private renderPrompts() {
        const { totalResults, noMatchesOverride } = this.props;
        const items = this.getFilteredData();
        if (this.state.loadingState === LoadingState.Complete && items.length === 0) {
            return (
                <ul
                    className={cx("autocomplete-prompt", "autocomplete-prompt--infoMsg")}
                    style={{
                        width: this.autocomplete ? this.autocomplete.clientWidth : "auto",
                    }}
                >
                    <li className="autocomplete-prompt-item">
                        {noMatchesOverride ?? this.t(TK.General_NoMatches)}
                    </li>
                </ul>
            )
        }
        const onAutocompleteSelection = (e: React.MouseEvent, { label, rawLabel, value, id }: IAutocompleteItem) => {
            e.stopPropagation();
            this.addItem({ label, rawLabel, value, id, isNew: false });
        }
        const hasMoreResults = totalResults ? totalResults >= 10 && items.length < totalResults : false;
        return items.length === 0 ?
            <div /> :
            (
                <ul className="autocomplete-prompt">
                    {items.map(item => (
                        <li
                            key={item.value}
                            className={"autocomplete-prompt-item"}
                            onClick={e => onAutocompleteSelection(e, item)}
                        >
                            {item.label}
                        </li>
                    ))}
                    {hasMoreResults && (
                        <li key="more-results" className={"autocomplete-prompt-item--more-results"}>
                            {this.t(TK.General_MoreResultsRefineQuery, { count: totalResults - items.length })}
                        </li>
                    )}
                </ul>
            );
    }

    /* HELPER METHODS */
    private onRequestAdd(e) {
        if (e.key.toUpperCase() === "ENTER") {
            e.preventDefault();
            this.onEnter(e);
        }
    }

    private addItem({ label, rawLabel, value, id, isNew = false }) {
        this.props.onAddNewClick({ label, rawLabel, value, id, isNew });
        this.setState({
            searchTerm: "",
            loadingState: LoadingState.Initial,
        });
        setTimeout(() => this.autocomplete.scrollTop = this.autocomplete.scrollHeight, 100);
    }

    /* EVENT HANDLERS */

    private onDeleteChip(value, label, isNew) {
        return () => this.props.onDeleteChipClick({ isNew, label, value });
    }

    private onEnter(e) {
        // todo scroll to bottom
        const searchTerm = e.target.value;
        if (searchTerm) {
            const proposals = this.getFilteredData();
            if (proposals.length) {
                this.addItem(proposals[0]);
                return;
            }
            if (this.props.allowItemCreation) {
                this.addItem(this.getEmptyPossibleAnswer(searchTerm));
            }
        }
    }

    private onBlur(e) {
        if (this.props.saveOnBlur) {
            const filteredItems = this.getFilteredData();
            if (!filteredItems.length) { // only save on blur if there is no match
                this.onEnter(e);
            }
        }
    }

    private getEmptyPossibleAnswer(searchTerm: string): IAutocompleteItem {
        return {
            id: "",
            isNew: true,
            label: searchTerm,
            rawLabel: searchTerm,
            value: searchTerm,
        };
    }

    private getFilteredData(): IAutocompleteItem[] {
        const { searchTerm } = this.state;
        if (searchTerm.length < 3) {
            return [];
        }
        const { data, selectedItems } = this.props;
        return data
            .filter(({ value }) => ( // filter out items that are already selected
                !selectedItems.some(({ value: selectedValue }) => selectedValue === value)
            ))
            .filter(item => ( // filter out items that don't match search term
                item.label.toLowerCase().indexOf(searchTerm.toLowerCase()) >= 0
            ));
    }
}

export default withTranslation()(Autocomplete);
