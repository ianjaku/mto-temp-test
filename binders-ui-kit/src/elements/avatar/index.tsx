import * as React from "react";
import Avatar from "@material-ui/core/Avatar";

export interface IAvatarProps {
    image: string;
    size?: number;
    // eslint-disable-next-line @typescript-eslint/ban-types
    style?: object;
}

const avatar: React.StatelessComponent<IAvatarProps> = props => (
    <Avatar
        src={props.image}
        style={{ ...props.style, width: `${props.size}px`, height: `${props.size}px` }}
    />
);

export default avatar;





