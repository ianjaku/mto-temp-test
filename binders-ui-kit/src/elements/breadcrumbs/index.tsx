import * as React from "react";
import BreadcrumbItem from "./item";
import CaretRight from "../icons/CaretRight";
import { Link } from "react-router-dom";
import debounce from "lodash.debounce";
import "./breadcrumbs.styl";

export interface IBreadcrumbItem {
    name: string;
    link?: string;
    readonly?: boolean;
    strikeThrough?: boolean;
    tooltip?: string;
    renderAsLast?: boolean;
}

export interface IBreadcrumbsProps {
    items: IBreadcrumbItem[];
    onClick?: (object) => void;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    itemContextMenu?: ((data) => React.ReactElement<any>) | JSX.Element;
    keySuffix?: string;
}

export interface IBreadcrumbState {
    breadcrumbsEl?: HTMLElement;
    itemsWrapperEl?: HTMLElement;
    hiddenElementsCount: number;
    truncatedElementsCount: number;
    lastClipOperation?: ClipOperation;
}

enum ClipOperation {
    Truncate,
    Hide,
}

class Breadcrumbs extends React.Component<IBreadcrumbsProps, IBreadcrumbState> {
    private onResize: () => void;

    // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
    constructor(props) {
        super(props);
        this.onResize = debounce(this.reMaybeClipBreadcrumbs.bind(this), 500);
        window.addEventListener("resize", this.onResize);
        this.maybeClipBreadcrumbs = this.maybeClipBreadcrumbs.bind(this);
        this.setItemsWrapperRef = this.setItemsWrapperRef.bind(this);
        this.setBreadcrumbsElRef = this.setBreadcrumbsElRef.bind(this);
        this.state = {
            hiddenElementsCount: 0,
            truncatedElementsCount: 0,
        };
    }

    public componentWillUnmount(): void {
        window.removeEventListener("resize", this.onResize);
    }

    // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
    public componentDidUpdate(_prevProps, prevState): void {
        const { itemsWrapper: prevItemsWrapper, breadcrumbsEl: prevBreadcrumbsEl } = prevState;
        const { itemsWrapperEl, breadcrumbsEl } = this.state;
        const breadcrumbsElChanged = breadcrumbsEl !== prevBreadcrumbsEl;
        const itemsWrapperElChanged = itemsWrapperEl !== prevItemsWrapper;
        if (breadcrumbsElChanged || itemsWrapperElChanged) {
            this.maybeClipBreadcrumbs();
        }
    }

    private maybeClipBreadcrumbs() {
        const { hiddenElementsCount, truncatedElementsCount, itemsWrapperEl, breadcrumbsEl, lastClipOperation } = this.state;
        let stateUpdate;
        let newLastClipOperation;
        if (lastClipOperation === undefined || (lastClipOperation === ClipOperation.Hide)) {
            stateUpdate = {
                truncatedElementsCount: truncatedElementsCount + 1,
            }
            newLastClipOperation = ClipOperation.Truncate;
        } else {
            stateUpdate = {
                hiddenElementsCount: hiddenElementsCount + 1,
            }
            newLastClipOperation = ClipOperation.Hide;
        }
        if (itemsWrapperEl.clientWidth > breadcrumbsEl.clientWidth) {
            this.setState({
                lastClipOperation: newLastClipOperation,
                ...stateUpdate,
            });
        }
    }

    private reMaybeClipBreadcrumbs() {
        this.setState({
            hiddenElementsCount: 0,
            truncatedElementsCount: 0,
        }, () => this.maybeClipBreadcrumbs());
    }

    private setItemsWrapperRef(ref) {
        const { itemsWrapperEl } = this.state;
        if (itemsWrapperEl === undefined) {
            this.setState({ itemsWrapperEl: ref });
        }
    }

    private setBreadcrumbsElRef(ref) {
        const { breadcrumbsEl } = this.state;
        if (breadcrumbsEl === undefined) {
            this.setState({ breadcrumbsEl: ref });
        }
    }

    public render(): JSX.Element {
        return (
            <div className="breadcrumbs" ref={this.setBreadcrumbsElRef}>
                <div className="breadcrumbs-itemswrapper" ref={this.setItemsWrapperRef}>
                    {this.renderItems()}
                </div>
            </div>
        );
    }

    private renderItems() {
        const { items, onClick, itemContextMenu, keySuffix } = this.props;

        const { hiddenElementsCount, truncatedElementsCount } = this.state;

        const generalMaxWidth = this.calculateMaxBreadcrumbWidth();
        const hasItemContextMenu = !!itemContextMenu;

        const isFirstLinkEmpty = (items.at(0)?.name?.length ?? 0) === 0;

        return items
            .slice(hiddenElementsCount + (isFirstLinkEmpty ? 1 : 0))
            .reduce((acc, { name, link, readonly, strikeThrough, tooltip, renderAsLast }, index, arr) => {
                const isLast = renderAsLast !== undefined ? renderAsLast : (arr.length === index + 1);
                const previousLink = index > 0 && items[index - 1].link;
                const isOffset = hiddenElementsCount > 0;
                const isEllipsis = isOffset && index === 0;

                const maxWidth = index < truncatedElementsCount ? generalMaxWidth : undefined;

                const breadCrumbItem = (
                    <BreadcrumbItem
                        key={`item-${index}${keySuffix}`}
                        onClick={onClick}
                        name={isEllipsis ? "..." : name}
                        link={link}
                        isLast={isLast}
                        isReadonly={readonly}
                        isStrikeThrough={strikeThrough}
                        tooltip={tooltip}
                        maxWidth={maxWidth}
                        itemContextMenu={itemContextMenu}
                        hasItemContextMenu={hasItemContextMenu}
                    />
                );

                const arrow = (previousLink || link ?
                    (
                        <Link
                            className="breadcrumbs-arrow"
                            key={`carrret-${index}${keySuffix}`}
                            to={previousLink || link}
                            tabIndex={-1}
                        >
                            {{ ...CaretRight("") }}
                        </Link>
                    ) :
                    (
                        <div className="breadcrumbs-arrow">
                            {{ ...CaretRight("", {}, `carrret-${index}${keySuffix}`) }}
                        </div>
                    )
                );

                return [
                    ...acc,
                    index > 0 ? arrow : null,
                    breadCrumbItem,
                ];
            }, []);
    }

    private calculateMaxBreadcrumbWidth(): number {
        const { items } = this.props;
        const { breadcrumbsEl } = this.state;
        const { hiddenElementsCount } = this.state;
        const visibleItemsCount = items.length - hiddenElementsCount;
        return breadcrumbsEl && (breadcrumbsEl.clientWidth / visibleItemsCount) || undefined;
    }
}

export default Breadcrumbs;
