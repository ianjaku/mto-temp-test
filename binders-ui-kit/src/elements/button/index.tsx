import * as React from "react";
import CircularProgress from "../circularprogress";
import autobind from "class-autobind";
import cx from "classnames";
import { withTheme } from "@material-ui/core/styles";
import "./Button.styl";

interface IButtonProps {
    CTA?: boolean;
    // Uses the branding primary color in the CTA & secondary buttons
    branded?: boolean;
    className?: string;
    downloadableName?: string;
    hrefAnchor?: string;
    icon?: React.ReactElement;
    iconRight?: React.ReactElement;
    id?: string;
    inactiveWithLoader?: boolean;
    isEnabled?: boolean;
    mouseOverActiveOnDisabled?: boolean;
    onBlur?: () => void;
    onClick?: (e?: React.MouseEvent<HTMLElement>) => void;
    onMouseEnter?: (e) => void;
    onMouseLeave?: (e) => void;
    onMouseOver?: (e) => void;
    onKeyDown?: React.KeyboardEventHandler<HTMLElement>;
    role?: string;
    secondary?: boolean;
    style?: React.CSSProperties;
    tabIndex?: number;
    text: string;
    theme?: { palette: { primary: { main: string; dark: string; } } };
    wrapper?: (children: React.ReactNode) => React.ReactNode;
    borderless?: boolean;
    "aria-label"?: string;
    "data-testid"?: string;
}

interface IButtonState {
    hovered: boolean;
}

class Button extends React.Component<IButtonProps & { forwardedRef?: React.Ref<HTMLDivElement> }, IButtonState> {
    public static defaultProps: Partial<IButtonProps> = {
        CTA: false,
        inactiveWithLoader: false,
        isEnabled: true,
        secondary: false,
    };
    constructor(props: IButtonProps) {
        super(props);
        this.onClick = this.onClick.bind(this);
        this.onKeyDown = this.onKeyDown.bind(this);
        this.state = {
            hovered: false,
        };
        autobind(this);
    }

    public render() {
        const { text, isEnabled, secondary, CTA, inactiveWithLoader,
            hrefAnchor, downloadableName, className, icon, iconRight, id, wrapper, borderless } = this.props;
        const classNames = cx(
            "button",
            !isEnabled ? "button--disabled" : "",
            inactiveWithLoader ? "button--unhoverable" : "",
            secondary ? "button--secondary" : "",
            CTA ? "button--CTA" : "",
            className,
            { branded: this.props.branded === true },
            { "button--borderless": borderless },
        );

        const button = (
            <div
                ref={this.props.forwardedRef}
                tabIndex={this.props.tabIndex}
                onBlur={this.props.onBlur}
                className={classNames}
                style={this.getThemeVariables()}
                onMouseOver={this.onMouseOver}
                onMouseLeave={this.onMouseLeave}
                onMouseEnter={this.props.onMouseEnter}
                onClick={inactiveWithLoader ? undefined : this.onClick}
                onKeyDown={inactiveWithLoader ? undefined : this.onKeyDown}
                id={id}
                role={this.props.role}
                arial-label={this.props["aria-label"]}
                {...this.props["data-testid"] ? { "data-testid": this.props["data-testid"] } : {}}
            >
                {icon}
                {text}
                <Loader enabled={inactiveWithLoader} isSecondary={secondary} />
                {hrefAnchor ? <a href={hrefAnchor} className="button-link" download={downloadableName} /> : ""}
                {iconRight}
            </div>
        );
        return wrapper ? wrapper(button) : button;
    }


    private onMouseOver(e) {
        const { onMouseOver, isEnabled, mouseOverActiveOnDisabled } = this.props;
        if (!isEnabled && !mouseOverActiveOnDisabled) {
            return;
        }
        this.setState({ hovered: true });
        if (onMouseOver) {
            onMouseOver(e);
        }
    }

    private onMouseLeave(e) {
        const { onMouseLeave, isEnabled, mouseOverActiveOnDisabled } = this.props;
        if (!isEnabled && !mouseOverActiveOnDisabled) {
            return;
        }
        this.setState({ hovered: false });
        if (onMouseLeave) {
            onMouseLeave(e);
        }
    }

    private getThemeVariables(): React.CSSProperties {
        const { theme } = this.props;

        // These are css variables intended to be used in the Style tag
        return {
            "--mainColor": theme.palette?.primary?.main,
            "--mainColor-hover": theme.palette?.primary?.dark,
        } as React.CSSProperties;
    }

    private onClick(e?: React.MouseEvent<HTMLElement>) {
        const { isEnabled, onClick } = this.props;
        if (isEnabled && onClick) {
            onClick(e);
        }
    }

    private onKeyDown(e?: React.KeyboardEvent<HTMLElement>): void {
        if (this.props.isEnabled) {
            this.props.onKeyDown?.(e);
        }
    }
}

const Loader: React.FC<{
    enabled?: boolean;
    isSecondary?: boolean;
}> = ({ enabled = false, isSecondary = false }) => {
    if (!enabled) return null;
    return CircularProgress(
        cx("button-loader", { "button-loader--secondary": isSecondary }),
        { position: "absolute" }
    );
}

const ForwardedRefButton = React.forwardRef<HTMLDivElement, IButtonProps>((props, ref) => {
    return <Button {...props} forwardedRef={ref} />;
});

export default withTheme(ForwardedRefButton);
