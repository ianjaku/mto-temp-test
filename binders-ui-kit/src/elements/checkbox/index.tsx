import * as React from "react";
import CheckBoxIcon from "@material-ui/icons/CheckBox";
import CheckBoxOutlineBlankIcon from "@material-ui/icons/CheckBoxOutlineBlank";
import FormControlLabel from "@material-ui/core/FormControlLabel";
import MaterialCheckbox from "@material-ui/core/Checkbox";
import Tooltip from "@material-ui/core/Tooltip";

export interface ICheckboxState {
    checked: boolean;
}

class Checkbox extends React.Component<ICheckboxProps, ICheckboxState> {
    public static defaultProps: Partial<ICheckboxProps> = {
        checked: false,
        disabled: false,
        isControlled: false,
    };

    // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
    static getDerivedStateFromProps(nextProps, prevState) {
        const { checked } = prevState;
        const { isControlled } = nextProps;
        return isControlled && checked !== nextProps.checked ?
            { checked: nextProps.checked } :
            null;
    }

    // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
    constructor(props) {
        super(props);
        this.onChange = this.onChange.bind(this);
        this.renderMaterialCheckbox = this.renderMaterialCheckbox.bind(this);
        this.state = {
            checked: props.checked,
        };
    }

    public render(): JSX.Element {
        const { label, labelStyle, tooltip } = this.props;
        const component = !label ?
            this.renderMaterialCheckbox() :
            (
                <FormControlLabel
                    label={label}
                    control={this.renderMaterialCheckbox()}
                    style={labelStyle}
                    className="checkboxControlLabel"
                />
            );
        return tooltip ? <Tooltip title={tooltip} placement="bottom-start"><div>{component}</div></Tooltip> : component;
    }

    private renderMaterialCheckbox() {
        const { disabled, iconStyle, className, iconSize } = this.props;

        const { checked } = this.state;
        return (
            <MaterialCheckbox
                checked={checked}
                disabled={disabled}
                onChange={this.onChange}
                className={className}
                icon={<CheckBoxOutlineBlankIcon fontSize={iconSize || "small"} />}
                checkedIcon={<CheckBoxIcon style={iconStyle} fontSize={iconSize || "small"} />}
                color="primary"
            />
        );
    }

    private onChange(e) {
        const { onCheck } = this.props;
        const { checked } = e.target;

        this.setState({ checked }, () => {
            if (typeof onCheck === "function") {
                onCheck(checked);
            }
        });
    }

    // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
    componentDidUpdate(prevProps): void {
        if (prevProps.checked !== this.props.checked) {
            this.setState({
                checked: this.props.checked
            });
        }
    }

}

export interface ICheckboxProps {
    onChange?: (isChecked: boolean) => void;
    onCheck?: (isChecked: boolean) => void;
    label?: string;
    disabled?: boolean;
    checked?: boolean;
    // eslint-disable-next-line @typescript-eslint/ban-types
    iconStyle?: object;
    isControlled?: boolean;
    // eslint-disable-next-line @typescript-eslint/ban-types
    labelStyle?: object;
    onFocus?: (e) => void;
    // eslint-disable-next-line @typescript-eslint/ban-types
    style?: object;
    className?: string;
    tooltip?: string;
    iconSize?: "default" | "inherit" | "large" | "medium" | "small";
}

export default Checkbox;
