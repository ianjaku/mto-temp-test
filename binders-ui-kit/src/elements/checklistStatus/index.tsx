import * as React from "react";
import { detectTimeZone, fmtDate, fmtDateIso8601TZ } from "@binders/client/lib/util/date";
import ChecklistStatusButton from "../button/ChecklistStatusButton";
import { IChecklist } from "@binders/client/lib/clients/repositoryservice/v3/contract";
import { TranslationKeys as TK } from "@binders/client/lib/react/i18n/translations";
import cx from "classnames";
import { useTranslation } from "@binders/client/lib/react/i18n";
import "./ChecklistStatus.styl";

const { useCallback, useMemo } = React;

interface IChecklistStatusProps {
    checklist: IChecklist;
    onCheck: (checklistId: string, performed: boolean) => void;
    disabled?: boolean;
    isBlockingProgress?: boolean;
    mockChecklistAs?: boolean; // Change performed to be true/false independent of the actual checklist value
}

function ChecklistStatus(props: IChecklistStatusProps): React.JSX.Element {
    const { checklist, disabled, onCheck, isBlockingProgress } = props;
    const { performedHistory } = checklist;

    const performed = useMemo(() => {
        if (props.mockChecklistAs == null) return checklist.performed;
        return props.mockChecklistAs;
    }, [props.mockChecklistAs, checklist.performed]);

    const [
        lastPerformedDate,
        lastPerformedByUserName,
        prevLastPerformedDate,
        prevLastPerformedByUserName,
    ] = useMemo(() => {
        if (!(performedHistory || []).length) {
            return [];
        }
        const lastHistory = performedHistory.sort((h1, h2) => h1.lastPerformedDate < h2.lastPerformedDate ? 1 : -1)[0];
        let prevLastPerformedDate: Date | null = null;
        let prevLastPerformedByUserName: string | null = null;
        if (performedHistory.length > 1) {
            prevLastPerformedDate = performedHistory[1].lastPerformedDate;
            prevLastPerformedByUserName = performedHistory[1].lastPerformedByUserName;
        }
        return [
            lastHistory.lastPerformedDate ? new Date(lastHistory.lastPerformedDate) : null,
            lastHistory.lastPerformedByUserName,
            prevLastPerformedDate ? new Date(prevLastPerformedDate) : null,
            prevLastPerformedByUserName
        ];
    }, [performedHistory]);

    const { t } = useTranslation();
    const handleCheck = useCallback((e) => {
        if (!props.disabled) {
            onCheck(checklist.id, !checklist.performed);
        }
        e.stopPropagation();
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [checklist, onCheck]);

    function renderPerformedByInfo(rawDate: Date, who: string, isHistory?: boolean) {
        const date = fmtDateIso8601TZ(rawDate);
        const time = fmtDate(rawDate, "HH:mm", { timeZone: detectTimeZone() });
        return (
            <div className={cx("checklistStatus-performedBy", { "checklistStatus-performedBy--isHistory": isHistory })}>
                <label>
                    {`${t(isHistory ? TK.Reader_ChecklistTaskPerformedHistoryWhen : TK.Reader_ChecklistTaskPerformedWhen, { date, time })}`}
                </label>
                <label>
                    {`${t(TK.Reader_ChecklistTaskPerformedWho, { who })}`}
                </label>
            </div>
        );
    }

    return (
        <div
            className={cx(
                "checklistStatus",
                { "checklistStatus--isPerformed": performed },
                { "checklistStatus--isDisabled": disabled },
            )}
        >
            <ChecklistStatusButton
                shouldShake={isBlockingProgress}
                isPerformed={performed}
                onClick={handleCheck}
                disabled={props.disabled}
            />
            {lastPerformedDate && renderPerformedByInfo(lastPerformedDate, lastPerformedByUserName)}
            {prevLastPerformedDate && renderPerformedByInfo(prevLastPerformedDate, prevLastPerformedByUserName, true)}
        </div>
    )
}

export default ChecklistStatus;
