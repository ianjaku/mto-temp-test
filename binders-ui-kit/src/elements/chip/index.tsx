import * as React from "react";
import MaterialChip from "@material-ui/core/Chip";

export interface IChipProps {
    onClick?: (e?: React.MouseEvent<HTMLElement>) => void;
    onDelete?: (e?: React.MouseEvent<HTMLElement>) => void;
    label: React.ReactNode;
    style?: React.CSSProperties;
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export default (props: IChipProps) => (
    <MaterialChip
        onClick={props.onClick}
        onDelete={props.onDelete}
        label={props.label}
        style={props.style}
    />
);


