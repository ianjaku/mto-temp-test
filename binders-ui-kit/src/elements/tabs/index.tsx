export { Tabs } from "./Tabs";
export { Pane } from "./Pane";