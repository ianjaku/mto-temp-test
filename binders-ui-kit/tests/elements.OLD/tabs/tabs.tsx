import * as React from "react";
import { Pane, Tabs } from "../../../src/elements/tabs";
import { create } from "react-test-renderer";
import { mount } from "enzyme";

const activeTabIndex = 1;

const titles = [
    "Title #1",
    "Title #2",
    "Title #3",
];

// eslint-disable-next-line @typescript-eslint/no-unused-vars
const createTabs = (initialIndex) => (
    <Tabs initialSelectedIndex={activeTabIndex} >
        <Pane label={titles[0]} />
        <Pane label={titles[1]} />
        <Pane label={titles[2]} />
        <Pane label={titles[3]} />
    </Tabs>
);

const createTabsWithoutInitializedActiveIndex = () => (
    <Tabs>
        <Pane label={titles[0]} />
        <Pane label={titles[1]} />
        <Pane label={titles[2]} />
        <Pane label={titles[3]} />
    </Tabs>
);


describe("Tabs", () => {

    test("Base tabs (snapshot)", () => {
        const tabs = create(createTabs(0));
        const serialized = tabs.toJSON();
        expect(serialized).toMatchSnapshot();
    });

    test("Switch tab (mount: enzyme+jsdom)", () => {
        const tabsMount = mount(createTabs(2));
        tabsMount.find(".tabs-nav .tabs-item").first().simulate("click");
        const isActiveTab = tabsMount.find(".tabs-nav .tabs-item").first();
        expect(isActiveTab.hasClass("active")).toEqual(true);
    });

    test("Tabs without initialized active index (mount: enzyme+jsdom)", () => {
        const tabsMount = mount(createTabsWithoutInitializedActiveIndex());
        const isActiveTab = tabsMount.find(".tabs-nav .tabs-item").first();
        expect(isActiveTab.hasClass("active")).toEqual(true);
    });

    test("Let's check tab content existing (mount: enzyme+jsdom)", () => {
        const tabsMount = mount(createTabs(0));
        tabsMount.find(".tabs-nav .tabs-item").first().simulate("click");
        const isContentTabActive = tabsMount.find(".tabs-content .tabs-pane").first();
        expect(isContentTabActive.hasClass("active")).toEqual(true);
        expect(isContentTabActive).toHaveLength(1);
    });

});
