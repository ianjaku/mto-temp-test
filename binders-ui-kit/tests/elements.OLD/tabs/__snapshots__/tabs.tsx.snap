// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs Base tabs (snapshot) 1`] = `
<section
  className="tabs"
>
  <section
    className="tabs-nav"
  >
    <ul
      className="tabs-nav-list"
    >
      <li
        className="tabs-item"
        onClick={[Function]}
      >
        <span
          className="tabs-title"
        >
          Title #1
        </span>
      </li>
      <li
        className="tabs-item active"
        onClick={[Function]}
      >
        <span
          className="tabs-title"
        >
          Title #2
        </span>
      </li>
      <li
        className="tabs-item"
        onClick={[Function]}
      >
        <span
          className="tabs-title"
        >
          Title #3
        </span>
      </li>
      <li
        className="tabs-item"
        onClick={[Function]}
      >
        <span
          className="tabs-title"
        />
      </li>
    </ul>
  </section>
  <div
    className="tabs-content"
  >
    <div
      className="tabs-pane fade"
    />
    <div
      className="tabs-pane fade active"
    />
    <div
      className="tabs-pane fade"
    />
    <div
      className="tabs-pane fade"
    />
  </div>
</section>
`;
